### ISSUE TITLE
How to bin integral vs floating point correctly (precision issue and general question)

### ISSUE THREAD

**Issue by emmenlau** [boostorg/histogram#336]
Dear @HDembinski and other authors, thanks for this awesome library. We have just started using it and the speed and versatility are impressive! We have a problem that manifests as a precision problem, but also shows a more general question.

The problem: We use the floating precision histogram to bin any kind of data. When we bin data (gray value camera images) of integer precision, we set lower and upper bound based on the data range to `[min, max+1]` and the number of bins to `upper bound - lower bound`. However there are sometimes cases where the numeric precision then leads to an empty last bin. This is due to rounding error, the values fall into `max-1` instead of `max`. We checked and the bin index is then computed as `(max-1).999...` which finally resolves to `max-1`. Is there a good solution to this? It leads to confusion that the last bin is empty, and its also a bit problematic in writing tests.

The underlying question: We are not completely certain what is a "recommended" way to set histogram bounds for integral data space. Are there common guidelines? In floating point precision it is more understandable what data range is covered by each bin. But in integral space, it seems the value `0` could be best represented by bin `[-0.5, 0.5)` rather than the more obvious `[0, 1)`. And help, links or pointers to documentation would be highly appreciated!

**Comment by HDembinski** [comment]
Thanks for the kind praise!

This is a recurring question and the standard answer is to use an `integer` axis for images. The `regular` axis is intended for real-valued input values from a continuous distribution, not for integer values.

So would using `boost::histogram::axis::integer` work for you? The integer axis uses exact integer arithmetic, and it is a bit faster than the `regular` axis, too.

If for some reason your input values are really floating point, you could add half the bin width to avoid the rounding errors. This is not an elegant solution, but it works.

This is also not an elegant solution, but you can even make your own axis class that does this step automatically. See https://www.boost.org/doc/libs/develop/libs/histogram/doc/html/histogram/guide.html#histogram.guide.expert.user_defined_axes for examples.

Basically, you would override the`index` method of the regular axis, just like in the first example and add half the bin width to the input.

**Comment by HDembinski** [comment]
Come to think of it, you could also override the `index` method with something that rounds to the nearest integer.

**Comment by HDembinski** [comment]
The correct solution for this is for us to support `boost::rational` as the value type of the regular axis, see https://www.boost.org/doc/libs/1_77_0/libs/rational/rational.html, which would enable exact rational arithmetic. This does not work currently, but I can look into it.

**Comment by emmenlau** [comment]
Thanks a lot @HDembinski for the help!

What you say makes sense. We just looked at the integer axis, and it most likely would help. But I understand it can not be used with arbitrary bin counts, correct? This is not a huge impediment but it would be a bit sad.

Also the suggestion to add half a bin width would seem feasible and possibly a good idea! But then I guess we could not use the range-based `fill()` anymore which gives a very noticeable speedup?

Support for `boost::rational` sounds awesome, but would it introduce a significant slow-down?

**Comment by emmenlau** [comment]
Can I also ask a related question, in case you know: from a completely different point of view, is it sensible or common to shift the bin bounds by half a bin width down? We where considering this idea that for example bin `0` would have bounds `[-0.5, 0.5)` (for the example bins of size one). We see some potential benefits that (a) the integer or floating value falls in the center of the bin rather than the border, and (b) it compensates for aforementioned rounding issues. But we do not feel confident about this route if it is a very uncommon thing to do...

**Comment by HDembinski** [comment]
> Also the suggestion to add half a bin width would seem feasible and possibly a good idea! But then I guess we could not use the range-based fill() anymore which gives a very noticeable speedup?

If you make your own axis which implements this hack, it will work with range-based fill().

> Support for boost::rational sounds awesome, but would it introduce a significant slow-down?

I don't know, I don't think the speed penalty would be terrible.

Yet another option that I forgot is to use the `axis::variable`, which you can construct in such a way that the bin edges are exactly where you want them to be. `axis::variable` is considerably slower than `axis::regular`, though.

**Comment by emmenlau** [comment]
Thanks a million for this fast and very helpful reply. I need to digest this new input a bit. Support for `boost::rational` sounds awesome, but your input has triggered new thoughts that require a bit more discussion on our side...

**Comment by HDembinski** [comment]
Once again thinking more about it and reading your original message again, is it correct that you only have an issue with the very last bin edge? Because then your need may be resolved if we implement #319

**Comment by emmenlau** [comment]
Sorry for the delay in feedback. We've just tested #344 and it does solve a corner case for us, thanks for this. However the original problem still remains because we commonly do use overflow bins.

> ... is it correct that you only have an issue with the very last bin edge?

We have not observed the problem with other bins yet. But I am under the impression that this is just by chance. The problem should happen with any bin, when just by chance the raw index position falls by a margin below the lower bound of the "correct" bin.

We've discussed this internally and can't make up our minds between the multiple possible workarounds. What I see so far:
 - using boost::rational: This seems very powerful and would certainly address the problem in a clean way. But it seems this also makes significant work in the implementation, and we could not find any benchmarks on boost::rational that would help to estimate the potential performance hit in boost::histogram. It would be sad if this "clean" solution incurs a significant performance overhead.
 - adding an epsilon of half a bin width in the computation of the index: This seems the easiest solution for us that should solve the problem in a generic way. I haven't tested, but the performance hit may also be negligible compared to the division. But its not entirely clear to me how this would be implemented in a clean way. If boost::histogram would handle integer binning internally with epsilon, and floating point binning without epsilon, it may add an ambiguity where the same value in different types is binned differently. On the other hand, if this is not handled inside boost::histogram, then how could we continue to use the range-based fill methods (that give the significant speed boost)?

**Comment by emmenlau** [comment]
After a longer discussion we are considering a third option, that I wanted to get your feedback on. The best of all worlds for us seems to be an automatically switching axis that supports integer and floating point. It should behave like the integer axis as long as the bounds and the scale are precisely integral values. In this case it correctly bins integer and floating point values with (fast) integer arithmetics. Only when any of the relevant parameters (bounds or scale) is floating point, it automatically switches to floating point arithmetics.

This seems to bring a set of nice benefits, and we could not see any downsides (yet). Benefits:
 - A single histogram instance gives the "perfect" integer arithmetics precision when legal, but also supports floating point arithmetics when needed. No need for the user to decide upfront, and significantly reduced risk of rounding errors.
 - In many cases this can give superior speed even for floating point histograms just as long as their parameters coincide with integral parameters.

Did we overlook some obstacles here? What do you think?

**Comment by emmenlau** [comment]
Hi @HDembinski , sorry to bother you with this question again, but it would be great to get your feedback on the above implementation idea. We've implemented such an axis as a test (an axis that supports both integer and floating point precision, and automatically switches between the two, depending on the user parameters).
Does that seem reasonable to you? And would that be something that boost::histogram could support native (rather that our test implementation)?

**Comment by HDembinski** [comment]
Sorry for not responding earlier. I think as a practical fix, it is fine to just shift the bins by half a bin width  if you are still considering this.

The axis that you mention, which uses exact integer arithmetic when everything is integral: this is already how floating point numbers with integral values work to my knowledge. Could you perhaps produce a small code example to show the wrong behavior and the expected?

**Comment by HDembinski** [comment]
In any case, if you can make a prototype for such an axis with these properties, I am happy to review and build a PR to include it.

That being said, I still think that the clean solution is to support boost::rational in the regular axis. I don't think that the performance hit is terrible, but we have to benchmark it.

Thanks for your endurance in trying to work out a solution.

**Comment by HDembinski** [comment]
Prototype of an axis that works with boost::rational:
https://godbolt.org/z/3af1doM9a

**Comment by emmenlau** [comment]
Awesome! I'll take a look at it in the evening, thanks a lot already!

**Comment by HDembinski** [comment]
@emmenlau Could you please post a minimal example that shows the bug you experiencing? What I said earlier, that integer arithmetic with doubles is exact to my knowledge, is only true when no fractions are computed in intermediate results. I would like to understand your issue better. Consider these two mathematically identical calculations of an index from a value:
```
index1 = ((x - a) * n) / (b - a)
index2 = (x - a) / ((b - a) / n)
```
These two calculations, when calculated with floating point types, are not giving the exact same result even when x, a, b, n are integer values. In my understanding, however, at least the first index1 should give a result identical to a calculation with integer types. I need to test this, though.

**Comment by HDembinski** [comment]
axis::regular currently uses yet another version of this calculation:
```
index3 = n * ((x - a) / (b - a))
```
https://github.com/boostorg/histogram/blob/bf7712f0d7b548f5f29487a18aa1ae9f221e08ad/include/boost/histogram/axis/regular.hpp#L330

This calculation also uses non-integer values as intermediate results, so it is not expected to give the exact same result as an integer calculation even when x, a, b, n are integer values.

**Comment by HDembinski** [comment]
For reference, numpy.histogram uses a rather expensive algorithm to handle values which end up exactly on bin edges:
https://github.com/numpy/numpy/blob/e2d35064df262efa6eb7dfe5bfc43160c73cf685/numpy/lib/histograms.py#L857

**Comment by HDembinski** [comment]
Here is a little numerical test. It looks like the calculation that axis::regular uses is the only one of the aforementioned alternatives that does not produce accurate results. The other calculations give the same result as a calculation with integer types.
```py
import numpy as np

rng = np.random.default_rng()

x = rng.integers(-1000000, 1000000, size=1000000)
a = rng.integers(-1000000, 1000000, size=1000000)
b = rng.integers(-1000000, 1000000, size=1000000)
n = b - a

x[x < a] = a[x < a]


# reference, uses integer calculation
index0 = (x - a) * n // (b - a)

index1 = ((x - a) * n / (b - a)).astype(int)

index2 = ((x - a) / ((b - a) / n)).astype(int)

index3 = (((x - a) / ((b - a))) * n).astype(int)

print("index1", np.sum(index0 != index1))
print("index2", np.sum(index0 != index2))
print("index3", np.sum(index0 != index3))
```

**Comment by HDembinski** [comment]
Ironically, another version of the calculation that seems to work is:
`((x - a) * (1.0 / ((b - a) / n)))`

**Comment by HDembinski** [comment]
That index2, index4 and index5 give the same answers is not a surprise, since by construction (b-a) / n == 1.0.

**Comment by vakokako** [comment]
@HDembinski thanks for all your inpute, here is an example of the problem we are experiencing: [godbolt](https://godbolt.org/z/oaPo6n17h).

**Comment by emmenlau** [comment]
@HDembinski your input and discussion is very enlightening, thanks! As @vakokako shows, the problem for us typically arises when using the axis with larger integral bounds, that cause small rounding errors. If a bin in the middle of the histogram is affected (as in @vakokako's example) the problem is not so visible. But in other, more common cases, the last bin can be affected, and it will remain completely empty. This causes typically some confusion with our users. Is this a suitable motivation for the issue?

When further looking into this, @vakokako found that we can use the more precise integral arithmetic when:
 - The difference `d` between upper and lower bound is (precisely) integral, and
 - The difference `d` is a multiple of the histogram size.

**Comment by vakokako** [comment]
Here is our solution axis to this: [godbolt](https://godbolt.org/z/f8fcrd6rd).

This is achieved by adding a bool flag `mIsIntegral` and changing the
computation of index if we now that the axis is `integral`, otherwise
forward to `regular` axis computation.

Example of axes where `mIsIntegral == true`:
- regular_precise(10, -5, 5)
- regular_precise(5, -5, 5)
- regular_precise(10, -5.5, 4.5) - shifting of bounds is also supported

Example of axes where `isIntegral() == false`:
- regular_precise(11, -5, 5)
- regular_precise(20, -5, 5)
- regular_precise(10, -5, 5.5)

The additional members for lower/upper bounds are not neccessary, they are there only cause we wanted to keep the dependency on `axis::regular` and it stores those as private members.

**Comment by Dominic-Stafford** [comment]
Hi, I was sent here from this issue by @henryiii from the above issue, and just wanted to ask if you anticipate being able to implement one of the fixes discussed here at some time soon? My issue just comes from integer values in the middle of the distribution falling on the wrong side of integer bin edges, so I think either the solution with boost::rational or a change in the calculation as you discussed should fix this.

**Comment by emmenlau** [comment]
Hi @Dominic-Stafford ! We have an implementation based on what you find in the godbolt-link above and it works quite well for us, all problems with precision are obsolete. However @HDembinski is much more knowledgeable about the possible pro's and con's with respect to i.e. the solution with `boost::rational`. What I can say is that we observe a tiny slowdown in benchmarks with our solution compared to the upstream code, which is surprising since we use integer math in the relevant case. I assume that `boost::rational` would be even slower, but again @HDembinski may know this much better.

**Comment by HDembinski** [comment]
Hi all, I am still confused why you cannot use an integer axis, because that should be faster than a regular axis and provides exact integer arithmetic.

**Comment by HDembinski** [comment]
> What I can say is that we observe a tiny slowdown in benchmarks with our solution compared to the upstream code, which is surprising since we use integer math in the relevant case. I assume that `boost::rational` would be even slower, but again @HDembinski may know this much better.

I also think that boost::rational would be slower, because one needs to spend more CPU cycles to do the same work. Your solution should be pretty fast since `mIsIntegral` is constant after construction and the branch predictors will take the right branch each time. That being said, predicting the performance of modern CPUs is very difficult even for experts. @emmenlau, why do you override `value_type value(real_index_type aIndex) const noexcept` your nice solution on Godbolt? Wouldn't the normal implementation work, too?

**Comment by emmenlau** [comment]
Thanks for the thumbs up! I should say that all the nice work was done by @vakokako .

> why do you override value_type value(real_index_type aIndex) const noexcept in your nice solution on Godbolt? Wouldn't the normal implementation work, too?

From a brief look at the code I agree that may not be required. Possibly it's just an artifact from a previous test. I'll check with @vakokako and get back to you today!

**Comment by HDembinski** [comment]
I did the check, it turns out that the implementation of value() also needs to be modified. The version that I use ensures that the upper edge of the last bin is equal to the upper edge that the user supplied, while the calculation that @vakokako only guarantees that for bin widths that are integral values.

**Comment by emmenlau** [comment]
Thanks a lot @HDembinski for the quick check, now I recall the same from our discussions!

Let me kindly ask, do you think something like this (our custom axis with implicit integral support) could be included into the official library? We would be more than happy to provide a reasonable PR. Or if you can do something based on our work, please feel free to copy and/or use the code at godbolt to your liking. Consider it to be public domain, or any license that suits your needs.

The axis is nice because at little extra code, and virtually no runtime overhead, it transparently supports the integral (precise) case but also the floating cases. No conscious action from the user required.

**Comment by HDembinski** [comment]
Yes, I plan to finally fix this and I will use your research. I am very grateful for the offer of a PR but I think I need to use this opportunity to also refactor regular which became quite complex with all the optional functionality. Good design is an iterative process which I can do more quickly on my own.

**Comment by emmenlau** [comment]
Thanks a lot for your consideration and help! I'm more than happy about this approach!

**Comment by Dominic-Stafford** [comment]
Thanks @HDembinski and @emmenlau! It would be really nice to have a solution in the official library since we're using this via the python bindings, but I agree it's worth taking the time to settle on a good solution.

**Comment by HDembinski** [comment]
Thanks for pinging me on this. I want to fix this, but I am not sure what is the best way.

**Comment by HDembinski** [comment]
The use case you have is covered by axis::integer<double>, I think, which should be faster than a regular axis with a special case for bins of integer width.

**Comment by ryanelandt** [comment]
I think the precision problems discussed in this issue may be addressed by PR draft #386. I added a [unit test](https://github.com/ryanelandt/histogram/blob/28838d2cf50982ed7546da9e86b77d9d810275b2/test/axis_int_resolver_test.cpp#LL108C1-L150C4) based on @vakokako's example that demonstrates exact `index` and `value` calculations for:
- integer bin widths (e.g., 1, 2, 3)
- fractional bin widths (e.g., 0.375, 1.625, 7.25)  
- various offsets (e.g., -0.5, 0.0, 1000.25).

This method is cheap, as it does **not** correct for numerical errors. Could this address some of the precision problems raised in this issue?

Note: this method is only exact when numbers take few floating point bits to represent. For example, it will be exact for a bin width of 1/2 = 0.5, but not for 1/3 = 0.333333333333333...
